contracts/oracle/src/contract/oracle/mod.rs

```diff
@@ -155,3 +155,214 @@ where
         assert_eq!(set.len(), subscribers.len());
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use currency::{lpn::Usdc, native::Nls, Currency as _};
+    use finance::{
+        coin::{Amount, Coin},
+        duration::Duration,
+        percent::Percent,
+        price,
+    };
+    use marketprice::{config::Config as PriceConfig, SpotPrice};
+    use sdk::cosmwasm_std::{
+        testing::{MockApi, MockQuerier, MockStorage},
+        Addr, DepsMut, Empty, QuerierWrapper, Storage, Timestamp,
+    };
+
+    use crate::{
+        alarms::Alarm,
+        contract::alarms::MarketAlarms,
+        state::{config::Config, supported_pairs::SupportedPairs},
+        swap_tree,
+    };
+
+    use super::{feed::Feeds, feeder::Feeders, Oracle};
+
+    type BaseCurrency = Usdc;
+
+    type NlsCoin = Coin<Nls>;
+    type UsdcCoin = Coin<Usdc>;
+
+    #[test]
+    fn test() {
+        const SAMPLE_PERIOD_SECS: u32 = 3;
+        const _: () = if SAMPLE_PERIOD_SECS < 3 {
+            panic!("Bug is reproduced with minimum of 3 seconds for the sample period.");
+        };
+
+        let mut storage: MockStorage = MockStorage::new();
+
+        let price_config: PriceConfig = PriceConfig::new(
+            Percent::HUNDRED,
+            Duration::from_secs(SAMPLE_PERIOD_SECS),
+            1,
+            Percent::HUNDRED,
+        );
+
+        let mut now_seconds: u64 = 1;
+
+        init(&mut storage, &price_config);
+
+        add_higher_alarm(&mut storage);
+
+        feed_below_price(&price_config, &mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        dispatch_1(&mut storage, now_seconds);
+
+        add_lower_alarm(&mut storage);
+
+        now_seconds += 1;
+
+        for _ in 0..SAMPLE_PERIOD_SECS.checked_add(1).unwrap() {
+            dispatch_0(&mut storage, now_seconds);
+
+            now_seconds += 1;
+        }
+
+        feed_normal_price(&price_config, &mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        dispatch_1(&mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        // Bug happens on this step.
+        dispatch_0(&mut storage, now_seconds);
+
+        feed_below_price(&price_config, &mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        dispatch_0(&mut storage, now_seconds);
```

Comment:
please structure it further ... now it is ~ 50 lines long and is difficult to read